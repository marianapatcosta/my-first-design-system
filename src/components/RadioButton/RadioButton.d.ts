import React, { CSSProperties, ChangeEvent } from 'react';
export interface RadioButtonProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Radio Button's label
     */
    label: string;
    /**
     * Radio Button's name
     */
    name?: string;
    /**
     * radio button' value
     */
    value: any;
    /**
     * Is Radio Button checked?
     */
    checked: boolean;
    /**
     * Is Radio Button disabled?
     */
    disabled?: boolean;
    /**
     * Change handler
     */
    onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}
export declare const RadioButton: React.FC<RadioButtonProps>;
