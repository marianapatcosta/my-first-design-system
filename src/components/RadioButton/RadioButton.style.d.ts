export declare const StyledRadioButtonWrapper: import("styled-components").StyledComponent<"label", any, {}, never>;
export declare const StyledRadioButton: import("styled-components").StyledComponent<"span", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledRadioButtonInput: import("styled-components").StyledComponent<"input", any, {}, never>;
