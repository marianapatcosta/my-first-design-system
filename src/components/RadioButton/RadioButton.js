var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyledRadioButtonWrapper, StyledRadioButton, StyledRadioButtonInput, } from './RadioButton.style';
export var RadioButton = function (_a) {
    var label = _a.label, id = _a.id, disabled = _a.disabled, checked = _a.checked, props = __rest(_a, ["label", "id", "disabled", "checked"]);
    return (_jsxs(StyledRadioButtonWrapper, __assign({ "aria-checked": checked }, { children: [_jsx(StyledRadioButtonInput, __assign({ "data-testid": id || 'radio-button', type: 'radio', role: 'radio', checked: checked, disabled: disabled }, props), void 0),
            _jsx(StyledRadioButton, { disabled: disabled }, void 0), label] }), void 0));
};
RadioButton.defaultProps = {
    id: '',
    className: '',
    style: {},
    name: '',
    value: '',
    label: '',
    checked: false,
    disabled: false,
    onChange: function () { return null; },
};
