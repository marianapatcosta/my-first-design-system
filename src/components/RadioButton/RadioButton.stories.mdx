import { Meta } from '@storybook/addon-docs/blocks';
import { RadioButton } from './RadioButton';

<Meta title='Design System|RadioButton' component={RadioButton} />

# RadioButton

Simple radio button component

## Usage

Use radio button component when it is intended to choose one value of a predefined group of options, in which each
option correspnds to a radio button.

- Default RadioButton:
  <Story name='Default'>
    <RadioButton label='option 1' />
  </Story>

## Other important notes

not applicable.
