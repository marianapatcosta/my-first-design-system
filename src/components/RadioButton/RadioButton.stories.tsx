import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { RadioButton, RadioButtonProps } from './RadioButton';

export default {
  title: 'RadioButton',
  component: RadioButton,
  args: {},
} as Meta;

const Template: Story<RadioButtonProps> = (args) => <RadioButton {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'option 1',
};
