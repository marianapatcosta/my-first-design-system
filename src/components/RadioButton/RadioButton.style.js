var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';
export var StyledRadioButtonWrapper = styled.label(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  display: flex;\n  align-items: center;\n  user-select: none;\n  font-size: 95%;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"], ["\n  ", "\n  display: flex;\n  align-items: center;\n  user-select: none;\n  font-size: 95%;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"])), baseStyles);
export var StyledRadioButton = styled.span(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: relative;\n  height: 0.8rem;\n  width: 0.8rem;\n  border: 0.063rem solid ", ";\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  cursor: pointer;\n  margin-right: 0.625rem;\n  margin-left: 0.125rem;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :after {\n    content: '';\n    height: 0;\n    width: 0;\n    background-color: ", ";\n    border-radius: 50%;\n    -webkit-border-radius: 50%;\n    -moz-border-radius: 50%;\n    -ms-border-radius: 50%;\n    -o-border-radius: 50%;\n    transition: all 0.3s ease-in-out;\n    -webkit-transition: all 0.3s ease-in-out;\n    -moz-transition: all 0.3s ease-in-out;\n    -ms-transition: all 0.3s ease-in-out;\n    -o-transition: all 0.3s ease-in-out;\n  }\n\n  ", "\n"], ["\n  position: relative;\n  height: 0.8rem;\n  width: 0.8rem;\n  border: 0.063rem solid ", ";\n  border-radius: 50%;\n  -webkit-border-radius: 50%;\n  -moz-border-radius: 50%;\n  -ms-border-radius: 50%;\n  -o-border-radius: 50%;\n  cursor: pointer;\n  margin-right: 0.625rem;\n  margin-left: 0.125rem;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :after {\n    content: '';\n    height: 0;\n    width: 0;\n    background-color: ", ";\n    border-radius: 50%;\n    -webkit-border-radius: 50%;\n    -moz-border-radius: 50%;\n    -ms-border-radius: 50%;\n    -o-border-radius: 50%;\n    transition: all 0.3s ease-in-out;\n    -webkit-transition: all 0.3s ease-in-out;\n    -moz-transition: all 0.3s ease-in-out;\n    -ms-transition: all 0.3s ease-in-out;\n    -o-transition: all 0.3s ease-in-out;\n  }\n\n  ", "\n"])), colors.highlight, colors.highlight, function (_a) {
    var disabled = _a.disabled;
    return disabled && "" + disabledStyles;
});
export var StyledRadioButtonInput = styled.input(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  position: absolute;\n  opacity: 0;\n  z-index: -1;\n\n  :checked + ", ":after {\n    width: 0.5rem;\n    height: 0.5rem;\n  }\n\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"], ["\n  position: absolute;\n  opacity: 0;\n  z-index: -1;\n\n  :checked + ", ":after {\n    width: 0.5rem;\n    height: 0.5rem;\n  }\n\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"])), StyledRadioButton, StyledRadioButton, colors.highlight, StyledRadioButton);
var templateObject_1, templateObject_2, templateObject_3;
