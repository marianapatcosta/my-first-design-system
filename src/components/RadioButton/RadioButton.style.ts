import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';

export const StyledRadioButtonWrapper = styled.label`
  ${baseStyles}
  display: flex;
  align-items: center;
  user-select: none;
  font-size: 95%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const StyledRadioButton = styled.span<{
  disabled?: boolean;
}>`
  position: relative;
  height: 0.8rem;
  width: 0.8rem;
  border: 0.063rem solid ${colors.highlight};
  border-radius: 50%;
  -webkit-border-radius: 50%;
  -moz-border-radius: 50%;
  -ms-border-radius: 50%;
  -o-border-radius: 50%;
  cursor: pointer;
  margin-right: 0.625rem;
  margin-left: 0.125rem;
  display: flex;
  justify-content: center;
  align-items: center;

  :hover {
    opacity: 0.75;
  }

  :after {
    content: '';
    height: 0;
    width: 0;
    background-color: ${colors.highlight};
    border-radius: 50%;
    -webkit-border-radius: 50%;
    -moz-border-radius: 50%;
    -ms-border-radius: 50%;
    -o-border-radius: 50%;
    transition: all 0.3s ease-in-out;
    -webkit-transition: all 0.3s ease-in-out;
    -moz-transition: all 0.3s ease-in-out;
    -ms-transition: all 0.3s ease-in-out;
    -o-transition: all 0.3s ease-in-out;
  }

  ${({ disabled }) => disabled && `${disabledStyles}`}
`;

export const StyledRadioButtonInput = styled.input`
  position: absolute;
  opacity: 0;
  z-index: -1;

  :checked + ${StyledRadioButton}:after {
    width: 0.5rem;
    height: 0.5rem;
  }

  :focus-visible + ${StyledRadioButton} {
    outline: 0.125rem solid ${colors.highlight};
  }

  :focus:not(:focus-visible) + ${StyledRadioButton} {
    outline: none;
  }
`;
