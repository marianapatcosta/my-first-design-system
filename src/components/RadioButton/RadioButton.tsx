import React, { CSSProperties, ChangeEvent } from 'react';
import {
  StyledRadioButtonWrapper,
  StyledRadioButton,
  StyledRadioButtonInput,
} from './RadioButton.style';

export interface RadioButtonProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Radio Button's label
   */
  label: string;
  /**
   * Radio Button's name
   */
  name?: string;
  /**
   * radio button' value
   */
  value: any;
  /**
   * Is Radio Button checked?
   */
  checked: boolean;
  /**
   * Is Radio Button disabled?
   */
  disabled?: boolean;
  /**
   * Change handler
   */
  onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}

export const RadioButton: React.FC<RadioButtonProps> = ({
  label,
  id,
  disabled,
  checked,
  ...props
}) => {
  return (
    <StyledRadioButtonWrapper aria-checked={checked}>
      <StyledRadioButtonInput
        data-testid={id || 'radio-button'}
        type='radio'
        checked={checked}
        disabled={disabled}
        {...props}
      />
      <StyledRadioButton disabled={disabled} />
      {label}
    </StyledRadioButtonWrapper>
  );
};

RadioButton.defaultProps = {
  id: '',
  className: '',
  style: {},
  name: '',
  value: '',
  label: '',
  checked: false,
  disabled: false,
  onChange: () => null,
};
