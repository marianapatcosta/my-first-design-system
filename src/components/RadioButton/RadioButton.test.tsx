import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { RadioButton } from './RadioButton';

describe('RadioButton', () => {
  const defaultProps = {
    label: 'option 1',
    checked: false,
    onChange: () => null,
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<RadioButton {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label', () => {
    const { getByText } = render(<RadioButton {...defaultProps} />);

    getByText('option 1');
  });

  test('should handle change events', () => {
    const onChangeMock = jest.fn();
    const { getByLabelText } = render(
      <RadioButton {...defaultProps} onChange={onChangeMock} />
    );
    const radioButton = getByLabelText(/option 1/i) as HTMLInputElement;
    fireEvent.click(radioButton);

    expect(onChangeMock).toBeCalled();
  });

  test('should return null on change if onChange prop is not passed', () => {
    const result =
      RadioButton.defaultProps?.onChange && RadioButton.defaultProps.onChange();
    expect(result).toBeNull();
  });

  describe('Disabled RadioButton', () => {
    test('should display label', () => {
      const { getByText } = render(
        <RadioButton {...defaultProps} disabled={true} />
      );

      getByText('option 1');
    });

    test('should be disabled', () => {
      const { getByTestId } = render(
        <RadioButton {...defaultProps} disabled />
      );
      const radioButton = getByTestId('radio-button');
      expect(radioButton).toHaveAttribute('disabled');
    });
  });
});
