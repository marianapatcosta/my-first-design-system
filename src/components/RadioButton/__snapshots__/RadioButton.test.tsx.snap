// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton renders without errors and matches snapshot 1`] = `
<label
  aria-checked="false"
  class="sc-bdfBwQ ivjLLh"
>
  <input
    class="sc-dlfnbm ccvpOj"
    data-testid="radio-button"
    name=""
    role="radio"
    type="radio"
    value=""
  />
  <span
    class="sc-gsTCUz uaMDM"
  />
  option 1
</label>
`;
